File: leetcode/exercises/3.1.6.medium.time_based_key_value_store.js
// -----------------------
// Created     : 11/12/2025
// Last Edited : 12/12/2025 
// Topics      : 
// Big O       :
// Problem Id  : 981. Time Based Key Value Store
// References  :
// Notes       :
// -----------------------

var TimeMap = function() {
    this.data = {};
};

/** 
 * @param {string} key 
 * @param {string} value 
 * @param {number} timestamp
 * @return {void}
 */
TimeMap.prototype.set = function(key, value, timestamp) {
    if (key in this.data) {
        this.data[key].push([value, timestamp]); 
    } else {
        this.data[key] = [[value, timestamp]];
    }
};

/** 
 * @param {string} key 
 * @param {number} timestamp
 * @return {string}
 */
TimeMap.prototype.get = function(key, timestamp) {

    if (!(key in this.data)) return "";

    let l = 0;
    let r = this.data[key].length - 1;
    let res = "";
    while (l <= r) {
        let m = Math.floor(l + (r - l) / 2);

        if (this.data[key][m][1] == timestamp){
            return this.data[key][m][0]
        }

        if (this.data[key][m][1] < timestamp){
            res = this.data[key][m][0];
            l = m + 1;
        } else {
            r = m - 1;
        }
    }
    return res
};

File: leetcode/exercises/3.1.6.medium.time_based_key_value_store.go
// -----------------------
// Created     : 13/12/2025
// Last Edited : 13/12/2025 
// Topics      : 
// Big O       :
// Problem Id  : 981. Time Based Key Value Store
// References  :
// Notes       :
// -----------------------

type Entry struct {
    value string
    timestamp int
}

type TimeMap struct {
    data map[string][]Entry
}


func Constructor() TimeMap {
    return TimeMap{
        data: make(map[string][]Entry),
    }
}


func (this *TimeMap) Set(key string, value string, timestamp int)  {
    this.data[key] = append(this.data[key], Entry{value: value, timestamp:timestamp})
}


func (this *TimeMap) Get(key string, timestamp int) string {
    size := len(this.data[key])
    res := ""

    if size == 0 { return res }

    l := 0
    r := size - 1
    m := 0
    for l <= r {
        m = l + (r - l) / 2
        if this.data[key][m].timestamp == timestamp { return this.data[key][m].value }

        if this.data[key][m].timestamp < timestamp {
            res = this.data[key][m].value
            l = m + 1
        } else {
            r = m - 1
        }
    }
    return res
}

